fix: dev target needs special permission, now VLLM version works for run.sh and Dev Container

keivenchang opened the pull request:
#### Overview:

Consolidate container dev targets and simplify privilege handling for improved developer experience and security.

#### Details:

- Merged `local-dev` and `dev` into single `dev` target across Dockerfile.vllm, build.sh, and run.sh
- Fixed environment variable paths (CARGO_HOME, RUSTUP_HOME) to use user home directory
- Added proper UID/GID mapping for ubuntu user in dev containers
- NOT ADDRESSED in this PR: Dockerfile.sglang, Dockerfile.tensorrt_llm, ...

#### Where should the reviewer start?

- container/run.sh - See simplified privilege handling and dev mode detection
- container/Dockerfile.vllm - Review the cleaned-up `dev` target and environment variable setup
- container/build.sh - Check default target changes

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

keivenchang committed:
Merge dev and local-dev into single target for run.sh and DevContainer

- Auto-detect local-dev images, enable ubuntu user with host UID/GID
- Single local-dev target works seamlessly in both run.sh and DevContainer
- Move environment variables to Dockerfile, use absolute paths
- Remove redundant options and duplicate mounts

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

Refactor HF_CACHE mounting and add WORKSPACE_DIR variable

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

fix: simplify image naming logic for consistent pattern

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

refactor: update container build and run scripts for improved dev experience

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

kthui commented on container/run.sh:
```diff
@@ -203,7 +205,11 @@ get_options() {
 
     if [ -z "$IMAGE" ]; then
-        IMAGE="dynamo:latest-${FRAMEWORK,,}"
+        if [[ "${FRAMEWORK^^}" == "VLLM" ]]; then
+            IMAGE="dynamo:latest-${FRAMEWORK,,}-local-dev"
+        else
+            IMAGE="dynamo:latest-${FRAMEWORK,,}"
+        fi
```
Can we remove the `*-local-dev`? Since all vllm will be tagged `dynamo:latest-vllm-local-dev`, which does not conflict with `dynamo:latest-trtllm` for example?

keivenchang replied:
Currently, you can build:
```
build.sh --target framework --framework VLLM
build.sh --target runtime --framework VLLM
build.sh --target dynamo_base --framework VLLM
```

These are all VLLM, and not `local-dev`

keivenchang replied:
Are you talking about line 208:
```
        if [[ "${FRAMEWORK^^}" == "VLLM" ]]; then
            IMAGE="dynamo:latest-${FRAMEWORK,,}-local-dev"
        else
            IMAGE="dynamo:latest-${FRAMEWORK,,}"
        fi
```

Or are you talking about line 280:
```
    # Auto-enable DEVCONTAINER_MODE if image contains "local-dev"
    if [[ "$IMAGE" == *"local-dev" ]]; then
        DEVCONTAINER_MODE=TRUE
        MOUNT_WORKSPACE=TRUE
        # Interactive mode is implied when MOUNT_WORKSPACE is TRUE
    fi
```

keivenchang replied:
1) First off, if I was here months ago, I'd tell everybody to stop using the word `latest` — I’ll clean this up in a future PR. Running commands against `latest` is neither reproducible nor idempotent. What “latest” means now is different from what it meant a minute ago, which makes debugging and reproducing issues painful. The correct convention is to tag images with a code hash + timestamp, which guarantees more consistent and and more reproducible results -- not 100% guaranteed, but still much better than `latest`.

2) There are multiple vLLM variants — vllm-local-dev, vllm-runtime, vllm-base, etc. Check the Dockerfiles with ... AS <target>. So, there isn’t just a single “vllm” version.

keivenchang commented on container/run.sh:
```diff
@@ -227,29 +233,11 @@ get_options() {
+    if [ -n "$HF_CACHE" ] && [ -z "$DEVCONTAINER_MODE" ]; then
         mkdir -p "$HF_CACHE"
         VOLUME_MOUNTS+=" -v $HF_CACHE:/root/.cache/huggingface"
     fi
```
Outdated.

keivenchang commented on container/run.sh:
```diff
@@ -324,8 +337,52 @@ error() {
+        # Dev Container-specific setup - the Dockerfile handles UID/GID mapping via build args
+        echo "Dev Container mode enabled - using ubuntu user with host UID/GID"
+        # Use ubuntu user (with correct UID/GID baked into image)
+        USER_ARGS+="--user ubuntu"
```
No, you should never run anything as root. You should always run as a user, for development.

keivenchang replied:
No, not true. People mount HF_CACHE and other custom stuff. Even .git or .ssh or whatever.

keivenchang commented on container/Dockerfile.vllm:
```diff
@@ -362,8 +364,24 @@ RUN uv pip install maturin[patchelf]
+ENV WORKSPACE_DIR=${WORKSPACE_DIR}
+
+# Make sure to sync this with the one specified on README.md
+ENV PYTHONPATH=${WORKSPACE_DIR}/components/metrics/src:${WORKSPACE_DIR}/components/frontend/src:${WORKSPACE_DIR}/components/planner/src:${WORKSPACE_DIR}/components/backends/mocker/src:${WORKSPACE_DIR}/components/backends/trtllm/src:${WORKSPACE_DIR}/components/backends/vllm/src:${WORKSPACE_DIR}/components/backends/sglang/src:${WORKSPACE_DIR}/components/backends/llama_cpp/src
```
This is a blanket PYTHONPATH specified in README.md, for all/any image. I don't want to get into this "for this image, do this, for that image, do that." It creates a maintenance nightmare.

keivenchang replied:
The README.md is agnostic to framework. So it doesn't make sense to remove that from the README.md. Will adding it hurt the vLLM runtime?

kthui commented on container/Dockerfile.vllm:
```diff
@@ -340,13 +340,15 @@ COPY --from=dynamo_base $CARGO_HOME $CARGO_HOME
     && echo "$USERNAME ALL=(root) NOPASSWD:ALL" > /etc/sudoers.d/$USERNAME \
     && chmod 0440 /etc/sudoers.d/$USERNAME \
     && mkdir -p /home/$USERNAME \
+    && groupmod -g $USER_GID $USERNAME \
```
what happen if you build the image on a compute node that has user id 1010 for example, and then run the container locally that only has user id 1000 and 1001?

kthui replied:
How do you handle the case if everybody need to compile on that image?

keivenchang replied:
The CI team also said, they don't want to publish dev images -- only runtime. The reason is, they take up too much space. Dev images are not built every day. They should only be needed when there's a change in cargo version, some big pip changes, or some big apt-get changes.

kthui replied:
> If you need it, you build it locally, and run it locally.

You see why I don't use any of the build.sh and run.sh 🙂

nnshah1 commented on container/run.sh:
```diff
@@ -324,8 +337,52 @@ error() {
+if [ -n "$MOUNT_WORKSPACE" ]; then
+    # Set HOME path based on container mode
+    if [ -n "$DEVCONTAINER_MODE" ]; then
+        HOME_PATH="/home/ubuntu"
```
can these be made the same?

keivenchang commented on container/Dockerfile.vllm:
```diff
@@ -340,13 +340,15 @@ COPY --from=dynamo_base $CARGO_HOME $CARGO_HOME
 ENV USERNAME=ubuntu
-ARG USER_UID=1000
-ARG USER_GID=1000
+ARG USER_UID
```
If you have no default and try to build, you will get an error, which is the right behavior.

I don't want people to default to 1000, and then 10 minutes later, complain that they can't write to a directory because they are not using the ID they're suppose to be using $(id -u).

kthui replied:
I'm not sure why matching the UID:GID to the host user worked? because I see the file that prevents compilation is owned by root, for example `/usr/local/cargo/git/db/mistral.rs-d7a5d833e16ad691`. (assuming your host user is not root?)

keivenchang committed:
refactor: simplify privileged mode handling in run.sh

- Consolidate USER_ARGS into PRIVILEGED_STRING variable
- PRIVILEGED_STRING now handles both --privileged and --user ubuntu flags
- Simplify logic for privilege and user settings

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: remove redundant PRIVILEGED variable assignment

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

nv-tusharma commented on container/build.sh:
```diff
@@ -466,7 +493,14 @@ fi
 
 if [[ $TARGET == "dev" ]]; then
-    BUILD_ARGS+=" --build-arg USER_UID=$(id -u) --build-arg USER_GID=$(id -g) "
+    # Use provided UID/GID or default to current user
```
Out of scope of this PR, but we will slowly start reducing our reliance on build.sh and run.sh for our docker containers. These build and run scripts will still exist but they will be limited to the scope of developer usag whereas CI will primarily rely on docker build alternatively.

keivenchang replied:
Sounds good. What do you plan to use for developers in the future?

nv-tusharma commented on container/run.sh:
```diff
@@ -324,8 +318,49 @@ error() {
+        PRIVILEGED_STRING+=" --user ubuntu"
+    else
+        # Standard workspace mode - enable privileged mode
+        # TODO(keivenc): Security risk, remove soon. Dockerfiles (trtllm, sglang) still need to run as root.
```
We are working on refactoring sglang and TRT-LLM, should be done this week :)
